src/gas_metering/mod.rs

```diff
@@ -130,13 +186,100 @@ impl Rules for ConstantCostRules {
 /// The above transformations are performed for every function body defined in the module. This
 /// function also rewrites all function indices references by code, table elements, etc., since
 /// the addition of an imported functions changes the indices of module-defined functions. If the
-/// the module has a NameSection, added by calling `parse_names`, the indices will also be updated.
+/// module has a NameSection, added by calling `parse_names`, the indices will also be updated.
 ///
 /// This routine runs in time linear in the size of the input module.
 ///
 /// The function fails if the module contains any operation forbidden by gas rule set, returning
 /// the original module as an Err.
-pub fn inject<R: Rules>(
+///
+/// ## Injected mutable global
+///
+/// The output module exports a mutable [i64] global with the specified name, which is used for
+/// tracking the gas left during execution. Overall mechanics are similar to the [Imported host
+/// function](#imported-host-function) method, aside from that a local injected gas counting
+/// function is called from each metering block intstead of an imported function, which should make
+/// the execution reasonably faster. Execution engine should take care of synchronizing the global
+/// with the runtime.
+pub fn inject<R: Rules, G: GasMeterable>(
+	input_module: G,
+	rules: &R,
+) -> Result<elements::Module, elements::Module> {
```

Comment:
After fixing the trait it would probably look like this:

```suggestion
pub fn inject<R: Rules, G: Backend>(
	input_module: Module,
	rules: &R,
	backend: G,
) -> Result<elements::Module, elements::Module> {
```